### ISSUE TITLE
allow runnableExamples in forward procs

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#309]
This issue came up here: https://github.com/nim-lang/Nim/pull/16484/files#r550295803, and prevents adding runnableExamples in procs like `cmpMem` which are forward declared without going through complex contorsions.

### Example 1
```nim
when true:
  proc main*() =
    ## comment
    runnableExamples: discard
  proc main() = discard # Error: redefinition of 'main'; previous declaration here: 
```

### Example 2
```nim
when true:
  proc main*()
    ## comment
    runnableExamples: discard # Error: invalid indentation
```

### Example 3
`defined(nimdoc)` has other issues, eg:
```nim
when true:
  when defined(nimdoc):
    proc main*(): int =
      ## comment
      runnableExamples: discard
  when not defined(nimdoc):
    proc main*(): int = result = 2
  static: doAssert main() == 2 # would fail
```

### Expected Output
There should be a simple way to add runnableExamples in a forward declared proc.

If we make Example 1 work, there's an inherent ambiguity, because a proc with no body (and an `=`) is valid and returns default value.

If we make Example 2 work, it'd require a parser change (which would need to be backported otherwise stdlib can't use this feature), and this may have other complications.

Example 3 creates more complications and is more complex to use.

### Proposed solution
Introduce a pragma `{.forward.}`

```nim
when true:
  proc main*() {.forward.} =
    ## comment
    runnableExamples: discard
  proc main() = discard # works
```

### Additional Information
https://github.com/nim-lang/RFCs/issues/163 is related

**Comment by disruptek** [comment]
Can you explain why the runnable example needs to appear on the predeclaration?

I'm struggling to understand the purpose of the contortions in the first place... 🤔

**Comment by timotheecour** [comment]
> Can you explain why the runnable example needs to appear on the predeclaration?

because putting it in the definition would be a noop, even with `--docInternal`
```nim
when true:
  proc main*()
    ## comment
  proc main() =
    runnableExamples: echo 12 # not run, not shown in docs
    discard
```

**Comment by juancarlospaco** [comment]
Use a RST `.. code-block::` block with the `:test:` thingy ?.

**Comment by timotheecour** [comment]
> Use a RST .. code-block:: block with the :test: thingy ?.

`code-block` is a relic from the past and needs to disappear:
* it's too easy to misuse (i can never remember the exact spacing to use) and doesn't give any warning/error when you misuse it (instead it silently wont' run the test)

* it has the wrong default, which is doing no compiling/running by default, and is often wrong, eg 
`nim doc -r lib/std/compilesettings.nim` doesn't complain even though `compileSettingSeq` is a typo:
```nim
proc querySettingSeq*(setting: MultipleValueSetting): seq[string] {.
  compileTime, noSideEffect.} = discard
  ## Can be used to get a multi-string compile-time option. Example:
  ##
  ## .. code-block:: Nim
  ##   const nimblePaths = compileSettingSeq(MultipleValueSetting.nimblePaths)
```
* it doesn't syntax highlight the code block
* it's slow (eg `nim rst2html doc/manual.rst` takes 18 seconds) because, unlike runnableExamples, there is no test joining
* it has many edge cases, eg if you don't set `nimcache` it fails with a cryptic:
```
nim doc /pathto/t11617.nim
Error: unhandled exception: Failed to create '/Users/timothee/git_clone/nim/timn/build/nimcache/t11617' [IOError]
```
* it doesn't have many of the features of runnableExamples (eg `--doccmd`)

Furthermore, there's exactly 0 examples in stdlib using `.. code-block::` with `:test:` at proc scope (again, because it's hard to remember how to use it)

That said, this does work:
```nim
when true: # works
  proc main*()
    ## comment 1
    ##
    ## .. code-block:: Nim
    ##    :test: "nim r $1"
    ##   doAssert 1+1 == 2, "zook1"
    ## comment 2
  proc main() = discard
```

but if you forget a newline before `code-block:: Nim` or if you use:
```nim
    ## .. code-block:: Nim
    ##   :test: "nim r $1"
```
it'll silently not run the test.

With this RFC you can instead write:
```nim
when true:
  proc main*() {.forward.} =
    ## comment 1
    runnableExamples:
      doAssert 1+1 == 2, "zook1"
    ## comment 2
  proc main() = discard
```

which is objectively better

**Comment by disruptek** [comment]
I didn't find your explanation very helpful, so I decided to modify your example until it matched my expectation of what a rational, thinking programmer might actually write:

```nim
proc main*()
proc main*() =
  ## comment
  runnableExamples:
    echo 12 # not run, not shown in docs
  echo 14
```

Note that there are a few differences from your example:
- the procedures share the same signature
- the procedures are not inside a `when` statement
- the comment appears inside the definition, as does the example
- the runnable example is generated by `nim doc`
- both forms of comment are successfully documented

Is there a reason we cannot merely fix the documentation generator or fix the documentation?  I guess that's really the question I was trying to ask in the first place, and I apologize if this was confusing.

**Comment by juancarlospaco** [comment]
Should `:test:` thingy be deprecated then?, looks very broken-ish by the description,
I can not even find much use of it on the Nim repo, nor nimble.

**Comment by xflywind** [comment]
> Should `:test:` thingy be deprecated then?, looks very broken-ish by the description,
> I can not even find any use of it on the Nim repo, nor nimble, nor anywhere. 🤔

see it in manual.rst and *.rst

**Comment by timotheecour** [comment]
>  decided to modify your example until it matched my expectation of what a rational, thinking programmer might actually write:

A rational, thinking programmer would want to stick to one definition rule to avoid duplication of doc comment and runnableExamples, which is what you'd end up with with your proposal if there are multiple implementations, eg:
```nim
proc main*()
when defined(js):
  proc main*() =
    ## comment
    runnableExamples: discard
    impl()
else:
  proc main*() =
    ## comment
    runnableExamples: discard
    impl()
```

Furthermore, your suggestion doesn't work with `memCmp`, which was one of the motivations for this RFC; here's a reduced example (the difference with your example being then `when` block):
`nim doc -r main` => won't run example, won't show doc comments
```nim
when not defined(js):
  proc main*()
  proc main*() =
    ## comment
    runnableExamples:
      echo 12 # not run, not shown in docs
    echo 14
```

whether or not this is a bug and can be fixed, the 1st problem is the bigger one.

**Comment by disruptek** [comment]
If you don't want to duplicate the comment and the example, wouldn't it be wiser to perform alternative implementations in a template?  Then you can be assured that you have a single comment and a single runnable example that works regardless of `when` gymnastics and correctly follows the `main` exported API symbol.

**Comment by timotheecour** [comment]
>  wouldn't it be wiser to perform alternative implementations in a template?

I don't understand what you're suggesting, please write a working example of how this would look like.

**Comment by disruptek** [comment]
```nim
when true:
  template mainImpl() =
     echo 3
else:
  template mainImpl() =
    echo 4
proc main*()
proc main*() =
  ## comment
  runnableExamples:
    main()
  mainImpl()
```

**Comment by alaviss** [comment]
@disruptek Your example doesn't scale for bigger procs, esp when generics come into play.

This feature is desirable for people developing cross-platform APIs, where implementations are spread out between multiple files.

**Comment by disruptek** [comment]
Am I the only one required to defend assertions?

**Comment by alaviss** [comment]
It's very simple, really.

```nim
when true:
  template mainImpl() = echo 3
else:
  template mainImpl() = echo 4
proc main*() =
  ## comment
  runnableExamples:
    main()
  mainImpl()
```

Declaring a prototype that refers to an implementation. It's almost as if that is the entire point of forward declaration.

And what happen when you start doing overloads?

```nim
type SQL = distinct string

when true:
  template fooImpl(a: string) = echo a

  template fooImpl(sql: SQL) =
     doSqlThings(sql)
else:
  template fooImpl(a: string) = echo repr a
  template fooImpl(sql: SQL) = echo string sql

proc foo*(a: string) =
  ## comment
  runnableExamples:
    foo("string")
  fooImpl(a)

proc foo*(sql: SQL) =
  ## comment
  runnableExamples:
    foo(SQL "drop table users;")
  fooImpl(sql)
```

Yea, it's basically forward declaration. Why would anyone ever want to do forward declaration with extra steps?

**Comment by disruptek** [comment]
Because it's not only what you want, it's also correct by construction.  There's no way to overlook an implementation, omit documentation, or have the runnable example (which is a test) not work.  Regardless of how many implementations you have, you will only ever have one test/doc per exported signature.

If you use your alternative, you'll find that it's at least the same amount of work but offers none of these guarantees of correctness.

**Comment by alaviss** [comment]
Isn't this RFC about fixing forward declaration so that you can put docs there and solve the point you're raising?

**Comment by disruptek** [comment]
No, my point is that the language doesn't need this extension.  I'm :-1: on changing the language with yet another form of syntax that is not required.

**Comment by Araq** [comment]
Not yet another pragma... We should get rid of forward declarations, once we have IC we'll make that happen.
